feat: implement object stores

yihuang opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Module parameters are loaded multiple times for each tx, although the raw bytes are cached, the decoding process is not.

Support an object store that is transient stores for `interface{}`, applications can cache objects directly, without doing unnecessary encoding/decoding.

Caching module parameters is an important use case for this, there are also many use cases out there.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

yihuang commented on CHANGELOG.md:
```diff
@@ -41,6 +41,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 ## Bug Fixes
 
 * (crypto) [#19691](https://github.com/cosmos/cosmos-sdk/pull/19745) Fix tx sign doesn't throw an error when incorrect Ledger is used.
+* (store) [#]() Support object store which is a transient store support `interface{}`.
```
```suggestion
* (store) [#206](https://github.com/crypto-org-chain/cosmos-sdk/pull/206) Support object store which is a transient store support `interface{}`.
```

mmsqe commented on store/types/store.go:
```diff
@@ -332,6 +365,7 @@ const (
+	StoreTypeObject
 )
 
 func (st StoreType) String() string {
```
miss `StoreTypeObject` here

mmsqe reviewed (commented):
need use `storetypes.GKVStore[V]` in [runtime](https://github.com/crypto-org-chain/cosmos-sdk/blob/6fbc72701fc7ae12970b5ce477e018fa310a26a2/runtime/store.go#L100) right?

mmsqe commented on store/types/store.go:
```diff
@@ -332,6 +365,7 @@ const (
+	StoreTypeObject
 )
 
 func (st StoreType) String() string {
```
miss `StoreTypeObject` here

yihuang replied:
fixed

yihuang commented:
> need use `storetypes.GKVStore[V]` in [runtime](https://github.com/crypto-org-chain/cosmos-sdk/blob/6fbc72701fc7ae12970b5ce477e018fa310a26a2/runtime/store.go#L100) right?

dependers to be changed in other PR.

Commit:
Support object store

generic interface

generic btree

generic cachekv

generic transient store

support ObjStore

changelog

Update CHANGELOG.md

Signed-off-by: yihuang <huang@crypto.com>

object store key

Apply review suggestions

fix merge conflict

fix snapshot

revert dependers

value validation

seems still need to replace store

object store in api

gas accounting for object store

Commit:
Revert "build(deps): Bump github.com/cosmos/iavl from 1.0.1 to 1.1.1 in store (#19770)"

This reverts commit ff30f6e0a4463f6de19678581e8d1114646cb1c7.

Commit:
Problem: store package has dependency issues

Solution:
- revert the iavl bump, v1.0.2 is the version referenced by outer packages

Commit:
fix interface

Commit:
Merge branch 'release/v0.50.x' into object-store

Signed-off-by: yihuang <huang@crypto.com>

yihuang commented on baseapp/baseapp.go:
```diff
@@ -344,6 +344,17 @@ func (app *BaseApp) MountMemoryStores(keys map[string]*storetypes.MemoryStoreKey
 	}
 }
 
+// MountObjectStores mounts all in-memory KVStores with the BaseApp's internal
```
```suggestion
// MountObjectStores mounts all transient object stores with the BaseApp's internal
```

Commit:
Update baseapp/baseapp.go

Signed-off-by: yihuang <huang@crypto.com>

Commit:
fix obj store

Commit:
fix test

Commit:
bump go-version in ci

Commit:
fix e2e tests

Commit:
gofump

Commit:
fix lint

Commit:
fix test
